pkg/strategy/grid/strategy.go

```diff
@@ -0,0 +1,274 @@
+	Symbol string `json:"symbol"`
+
+	Interval types.Interval `json:"interval"`
+
+	// GridPips is the pips of grid, e.g., 0.001
+	GridPips fixedpoint.Value `json:"gridPips"`
+
+	// GridNum is the grid number (order numbers)
+	GridNum int `json:"gridNumber"`
+
+	BaseQuantity float64 `json:"baseQuantity"`
+
+	activeBidOrders map[uint64]types.Order
+	activeAskOrders map[uint64]types.Order
+
+	boll *indicator.BOLL
+	mu   sync.Mutex
+}
+
+func (s *Strategy) Subscribe(session *bbgo.ExchangeSession) {
+	// currently we need the 1m kline to update the last close price and indicators
+	session.Subscribe(types.KLineChannel, s.Symbol, types.SubscribeOptions{Interval: s.Interval.String()})
+}
+
+func (s *Strategy) updateBidOrders(orderExecutor bbgo.OrderExecutor, session *bbgo.ExchangeSession) {
+	quoteCurrency := s.Market.QuoteCurrency
+	balances := session.Account.Balances()
+
+	balance, ok := balances[quoteCurrency]
+	if !ok || balance.Available <= 0.0 {
+		return
+	}
+
+	var numOrders = s.GridNum - len(s.activeBidOrders)
+	if numOrders <= 0 {
+		return
+	}
+
+	var downBand = s.boll.LastDownBand()
+	var startPrice = downBand
+
+	var submitOrders []types.SubmitOrder
+	for i := 0; i < numOrders; i++ {
+		submitOrders = append(submitOrders, types.SubmitOrder{
+			Symbol:      s.Symbol,
+			Side:        types.SideTypeBuy,
+			Type:        types.OrderTypeLimit,
+			Market:      s.Market,
+			Quantity:    s.BaseQuantity,
+			Price:       startPrice,
+			TimeInForce: "GTC",
+		})
+
+		startPrice -= s.GridPips.Float64()
+	}
+
+	orders, err := orderExecutor.SubmitOrders(context.Background(), submitOrders...)
+	if err != nil {
+		log.WithError(err).Error("submit bid order error")
+		return
+	}
+
+	s.mu.Lock()
+	for i := range orders {
+		var order = orders[i]
+		log.Infof("adding order %d to the active bid order pool...", order.OrderID)
+		s.activeBidOrders[order.OrderID] = order
+	}
+	s.mu.Unlock()
+}
+
+func (s *Strategy) updateAskOrders(orderExecutor bbgo.OrderExecutor, session *bbgo.ExchangeSession) {
+	baseCurrency := s.Market.BaseCurrency
+	balances := session.Account.Balances()
+
+	balance, ok := balances[baseCurrency]
+	if !ok || balance.Available <= 0.0 {
+		return
+	}
+
+	var numOrders = s.GridNum - len(s.activeAskOrders)
+	if numOrders <= 0 {
+		return
+	}
+
+	var upBand = s.boll.LastUpBand()
+	var startPrice = upBand
+
+	var submitOrders []types.SubmitOrder
+	for i := 0; i < numOrders; i++ {
+		submitOrders = append(submitOrders, types.SubmitOrder{
+			Symbol:      s.Symbol,
+			Side:        types.SideTypeSell,
+			Type:        types.OrderTypeLimit,
+			Market:      s.Market,
+			Quantity:    s.BaseQuantity,
+			Price:       startPrice,
+			TimeInForce: "GTC",
+		})
+
+		startPrice += s.GridPips.Float64()
+	}
+
+	orders, err := orderExecutor.SubmitOrders(context.Background(), submitOrders...)
+	if err != nil {
+		log.WithError(err).Error("submit ask order error")
+		return
+	}
+
+	s.mu.Lock()
+	for i := range orders {
```

Comment:
ditto

Reply:
it was because it's an iterator variable, i wrote this to avoid side effect